static/app/components/profiling/ThreadSelector.tsx

```diff
@@ -0,0 +1,220 @@
+    if (!a.duration) {
+      return 1;
+    }
+
+    if (a.name.startsWith('(tid') && b.name.startsWith('(tid')) {
+      return -1;
+    }
+    if (a.name.startsWith('(tid')) {
+      return -1;
+    }
+    if (b.name.startsWith('(tid')) {
+      return -1;
+    }
+    if (a.name.includes('main')) {
+      return -1;
+    }
+    if (b.name.includes('main')) {
+      return 1;
+    }
+    return a.name > b.name ? -1 : 1;
+  });
+};
+
+interface ThreadSelectorProps {
+  activeProfileIndex: ProfileGroup['activeProfileIndex'];
+  onProfileIndexChange: (index: number) => void;
+  profileGroup: ProfileGroup;
+}
+
+function ThreadMenuSelector(props: ThreadSelectorProps): React.ReactElement | null {
+  const containerRef = React.useRef<HTMLDivElement>(null);
+  const [open, setOpen] = React.useState<boolean>(false);
+
+  const handleSelectItem = React.useCallback(
+    (p: Fuse.FuseResultWithMatches<Profile & {index: number}>) => {
+      const index = props.profileGroup.profiles.findIndex(
+        profile => profile.name === p.item.name
+      );
+      if (index === -1) {
+        return;
+      }
+      props.onProfileIndexChange(index);
+    },
+    [props.onProfileIndexChange, props.profileGroup]
+  );
+
+  React.useEffect(() => {
+    const tagNamesToPreventStealingFocusFrom = new Set<Element['tagName']>([
+      'INPUT',
+      'TEXTAREA',
+    ]);
+    function handleKeyDown(evt) {
+      if (tagNamesToPreventStealingFocusFrom.has(evt.target.tagName)) {
+        return;
+      }
+      if (evt.key === 't' && !open) {
+        evt.preventDefault();
+        setOpen(true);
+      }
+      if (evt.key === 'Escape' && open) {
+        setOpen(false);
+      }
+    }
+
+    document.addEventListener('keydown', handleKeyDown);
+    return () => document.removeEventListener('keydown', handleKeyDown);
+  }, [open]);
+
+  useOnClickOutside(containerRef, () => {
+    setOpen(false);
+  });
+
+  return (
+    <div>
+      <Button size="zero" borderless onClick={() => setOpen(true)}>
+        {props.profileGroup.profiles[props.activeProfileIndex]?.name ?? 'Select Thread'}
+      </Button>
+      <AutoComplete
+        // @ts-ignore the type is typed as any
+        onSelect={handleSelectItem}
+        closeOnSelect={false}
+        isOpen={open}
+      >
+        {({getInputProps, getItemProps, isOpen, inputValue, highlightedIndex}) => {
+          const sortedProfiles = sortProfiles(props.profileGroup.profiles);
+          const filteredProfiles = inputValue
+            ? new Fuse(sortedProfiles, {includeMatches: true, keys: ['name']}).search(
+                inputValue
+              )
+            : sortedProfiles.map(p => ({item: p, matches: [], score: 1}));
+
+          return isOpen ? (
+            <ThreadSelectorContainer ref={containerRef}>
+              <Input
+                type="search"
+                autoFocus
+                {...getInputProps({
+                  onKeyDown: evt => {
+                    if (evt.key === 'Escape') {
```

Comment:
Does the key down listener on the document not handle this already?

Reply:
No, it never fires... I think the component is preventing it somewhere

Reply:
Ah sorry, my comment went in the wrong place. I couldnt comment on the approval message you left... Yes for sure, should we add this to our board or design UI team? Idk what's a good place to do that